File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        cols = collections.defaultdict(set)
        rows = collections.defaultdict(set)
        squares = collections.defaultdict(set)    # key = (row/3, col/3)

        for r in range(9):
            for c in range(9):
                # empty cell
                if board[r][c] == ".":
                    continue
                # duplicate
                if(board[r][c] in rows[r]
                    or board[r][c] in cols[c]
                    or board[r][c] in squares[(r//3, c//3)]):
                        return False
                
                rows[r].add(board[r][c])
                cols[c].add(board[r][c])
                squares[(r//3, c//3)].add(board[r][c])
        return True

File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_set<char> rows[9];
        unordered_set<char> cols[9];
        unordered_set<char> squares[3][3];

        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                char current = board[r][c];
                if (current == '.') continue;

                // duplicates
                if (rows[r].count(current) || cols[c].count(current) || squares[r / 3][c / 3].count(current)) {
                    return false;
                }

                rows[r].insert(current);
                cols[c].insert(current);
                squares[r / 3][c / 3].insert(current);
            }
        }
        return true;
    }
};